File: src/lib/modes/sql/dialects.ts
// SQL dialect registry — TypeScript mirror of `src-tauri/src/modes/sql/dialects.rs`.
//
// Single source of truth for per-driver metadata (display names, default
// ports, abbreviations, parser profiles, capability flags) consumed by the
// SQL UI. Keep the entries here in lockstep with the Rust descriptor list.

export type SqlDialectKey = 'postgresql' | 'mysql' | 'sqlite' | 'clickhouse' | 'd1';

export interface SqlDialectDescriptor {
  key: SqlDialectKey;
  displayName: string;
  abbreviation: string;
  defaultPort: number;
  usesHostPort: boolean;
  usesCredentials: boolean;
  /** Matches the CodeMirror lang-sql dialect name. */
  parserProfile: string | null;
  /** Quote character used for delimited identifiers in DDL emitted from
   * the UI (e.g. DROP DATABASE). MySQL uses backticks; the rest use
   * double quotes per ANSI SQL. */
  identifierQuote: string;
}

export const SQL_DIALECTS: readonly SqlDialectDescriptor[] = [
  { key: 'postgresql', displayName: 'PostgreSQL', abbreviation: 'PG', defaultPort: 5432, usesHostPort: true,  usesCredentials: true,  parserProfile: 'PostgreSQL', identifierQuote: '"' },
  { key: 'mysql',      displayName: 'MySQL',      abbreviation: 'MY', defaultPort: 3306, usesHostPort: true,  usesCredentials: true,  parserProfile: 'MySQL',      identifierQuote: '`' },
  { key: 'sqlite',     displayName: 'SQLite',     abbreviation: 'SL', defaultPort: 0,    usesHostPort: false, usesCredentials: false, parserProfile: 'SQLite',     identifierQuote: '"' },
  // ClickHouse is HTTP-based (default 8123). node-sql-parser has no
  // dedicated profile, so we pin PostgreSQL as the closest fallback;
  // ClickHouse uses backticks for identifier quoting in DDL.
  { key: 'clickhouse', displayName: 'ClickHouse', abbreviation: 'CH', defaultPort: 8123, usesHostPort: true,  usesCredentials: true,  parserProfile: 'PostgreSQL', identifierQuote: '`' },
  // Cloudflare D1 — HTTPS-only, SQLite-flavoured. usesHostPort + usesCredentials
  // are false so the generic host/port/user/pass form is hidden; the
  // ConnectionDialog renders D1-specific fields (Account ID / Database ID /
  // API Token) when driver === 'd1'.
  { key: 'd1',         displayName: 'Cloudflare D1', abbreviation: 'D1', defaultPort: 0, usesHostPort: false, usesCredentials: false, parserProfile: 'SQLite',     identifierQuote: '"' },
] as const;

export function descriptorFor(key: string): SqlDialectDescriptor | undefined {
  return SQL_DIALECTS.find((d) => d.key === key);
}

export function defaultPortFor(key: string): number {
  return descriptorFor(key)?.defaultPort ?? 0;
}

/** Returns the parser profile for the given driver key, falling back to
 * `'PostgreSQL'` for unknown / not-yet-registered drivers — same default
 * the editor used before the registry existed. */
export function parserProfileFor(key: string): string {
  return descriptorFor(key)?.parserProfile ?? 'PostgreSQL';
}

File: src-tauri/src/modes/sql/dialects.rs
// SQL dialect registry — single source of truth for the per-driver
// metadata that used to be scattered across `client.rs` and the SvelteKit
// layer (default ports, display names, abbreviations, parser profiles).
//
// Per-dialect runtime logic that genuinely differs between drivers
// (connection-string syntax in `client::build_connection_url`, sqlx pool
// construction in `client::build_pool_inner`, dialect-specific SQL in
// `ai_tools.rs`) intentionally stays as match arms — those branches live
// inside the engine, not the metadata. The registry's job is to centralise
// the metadata and let dispatch happen via `descriptor_for_key(...)`.

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SqlDialect {
    Postgres,
    MySql,
    Sqlite,
    Clickhouse,
    /// Cloudflare D1 — HTTPS-only serverless SQLite. Driver is
    /// SQLite-flavoured but transport-wise mirrors ClickHouse (reqwest +
    /// JSON envelope, no sqlx pool).
    D1,
}

/// Registry record; some metadata fields are consumed only by the
/// TypeScript mirror today (`src/lib/modes/sql/dialects.ts`) and are kept
/// here so the Rust side stays the source of truth as the registry grows.
#[allow(dead_code)]
#[derive(Debug)]
pub struct SqlDialectDescriptor {
    pub dialect: SqlDialect,
    /// Stable key persisted in the saved-connections table and exchanged
    /// over the Tauri IPC boundary.
    pub key: &'static str,
    pub display_name: &'static str,
    pub abbreviation: &'static str,
    pub default_port: u16,
    pub uses_host_port: bool,
    pub uses_credentials: bool,
    /// Matches the CodeMirror lang-sql dialect name on the frontend.
    pub frontend_parser_profile: Option<&'static str>,
}

const DIALECTS: &[SqlDialectDescriptor] = &[
    SqlDialectDescriptor {
        dialect: SqlDialect::Postgres,
        key: "postgresql",
        display_name: "PostgreSQL",
        abbreviation: "PG",
        default_port: 5432,
        uses_host_port: true,
        uses_credentials: true,
        frontend_parser_profile: Some("PostgreSQL"),
    },
    SqlDialectDescriptor {
        dialect: SqlDialect::MySql,
        key: "mysql",
        display_name: "MySQL",
        abbreviation: "MY",
        default_port: 3306,
        uses_host_port: true,
        uses_credentials: true,
        frontend_parser_profile: Some("MySQL"),
    },
    SqlDialectDescriptor {
        dialect: SqlDialect::Sqlite,
        key: "sqlite",
        display_name: "SQLite",
        abbreviation: "SL",
        default_port: 0,
        uses_host_port: false,
        uses_credentials: false,
        frontend_parser_profile: Some("SQLite"),
    },
    SqlDialectDescriptor {
        dialect: SqlDialect::Clickhouse,
        key: "clickhouse",
        display_name: "ClickHouse",
        abbreviation: "CH",
        default_port: 8123,
        uses_host_port: true,
        uses_credentials: true,
        // node-sql-parser has no ClickHouse profile; PostgreSQL is the
        // closest fallback (similar quoting/casting). The editor falls
        // back to PostgreSQL anyway for unknown profiles, but we make it
        // explicit here so the registry stays the source of truth.
        frontend_parser_profile: Some("PostgreSQL"),
    },
    SqlDialectDescriptor {
        dialect: SqlDialect::D1,
        key: "d1",
        display_name: "Cloudflare D1",
        abbreviation: "D1",
        default_port: 0,
        // D1 doesn't use host/port or user/pass — the connection dialog
        // surfaces Account ID / Database ID / API Token fields instead.
        // Both flags are false so the existing generic fields stay hidden.
        uses_host_port: false,
        uses_credentials: false,
        frontend_parser_profile: Some("SQLite"),
    },
];

pub fn descriptor_for_key(key: &str) -> Option<&'static SqlDialectDescriptor> {
    DIALECTS.iter().find(|d| d.key == key)
}

/// Lookup by enum variant. Currently unused at runtime — `client.rs`
/// dispatches via `descriptor_for_key` then matches on `dialect` directly —
/// but kept as part of the registry's public surface for future callers
/// (e.g. ai_tools migration, ClickHouse onboarding).
#[allow(dead_code)]
pub fn descriptor_for(dialect: SqlDialect) -> &'static SqlDialectDescriptor {
    // Safe: the registry always contains every variant of `SqlDialect`.
    DIALECTS
        .iter()
        .find(|d| d.dialect == dialect)
        .expect("SqlDialect variant missing from registry")
}

#[allow(dead_code)]
pub fn all_descriptors() -> &'static [SqlDialectDescriptor] {
    DIALECTS
}
